test/workbox-background-sync/node/lib/test-Queue.mjs

```diff
@@ -0,0 +1,535 @@
+
+
+let Queue;
+
+/**
+ * Suspends execution in an async function for the specified amount of time.
+ *
+ * @param {number} time
+ * @return {Promise}
+ */
+const sleep = (time) => new Promise((resolve) => setTimeout(resolve, time));
+
+
+const clearObjectStore = async () => {
+  const db = await indexedDBHelper.getDB(
+      'workbox-background-sync', 'requests', {autoIncrement: true});
+
+  const items = await db.getAll();
+  for (const [key] of items.entries()) {
+    await db.delete(key);
+  }
+};
+
+
+describe(`backgroundSync.Queue`, function() {
+  const sandbox = sinon.sandbox.create();
+
+  beforeEach(async function() {
+    sandbox.restore();
+    clearRequire.match(RegExp('workbox-background-sync/lib/Queue.mjs'));
+
+    clearObjectStore();
+
+    // Remove any lingering event listeners
+    global.__removeAllEventListeners();
+
+    // Re-import Queue each time so the name map gets reset.
+    const imprt = await import(
+        '../../../../packages/workbox-background-sync/lib/Queue.mjs');
+
+    Queue = imprt.default;
+  });
+
+  after(async function() {
+    sandbox.restore();
+    clearRequire.match(RegExp('workbox-background-sync/lib/Queue.mjs'));
+
+    clearObjectStore();
+
+    // Remove any lingering event listeners
+    global.__removeAllEventListeners();
+  });
+
+  describe(`constructor`, function() {
+    it(`should throw if two queues are created with the same name`,
+        async function() {
+      expect(() => {
+        new Queue('foo');
+        new Queue('bar');
+      }).not.to.throw();
+
+      await expectError(() => {
+        new Queue('foo');
+      }, 'duplicate-queue-name');
+
+      expect(() => {
+        new Queue('baz');
+      }).not.to.throw();
+    });
+
+    it(`should add a sync event listener that replays the queue when the ` +
+        `event is dispatched`, async function() {
+      sandbox.spy(self, 'addEventListener');
+      sandbox.stub(Queue.prototype, 'replayRequests');
+
+      new Queue('foo');
+
+      expect(self.addEventListener.calledOnce).to.be.true;
+      expect(self.addEventListener.calledWith('sync')).to.be.true;
+
+      self.dispatchEvent(new SyncEvent('sync', {
+        tag: 'workbox-background-sync:foo',
+      }));
+
+      expect(Queue.prototype.replayRequests.calledOnce).to.be.true;
+    });
+
+    it(`should try to replay the queue on SW startup in browsers that ` +
+        `don't support the sync event`, async function() {
+      // Delete the SyncManager interface to mock a non-supporting browser.
+      const originalSyncManager = registration.sync;
+      delete registration.sync;
+
+      sandbox.stub(Queue.prototype, 'replayRequests');
+
+      new Queue('foo');
+
+      expect(Queue.prototype.replayRequests.calledOnce).to.be.true;
+
+      registration.sync = originalSyncManager;
+    });
+  });
+
+  describe(`createPlugin`, function() {
+    it(`should return an object implementing the fetchDidFail plugin ` +
```

Comment:
This method still feels weird to me. May be naming, but I'm struggling to understand what it's for.